File: native/profilestate/state.go
// Package profilestate stores one bounded private document in the same protected
// SQLite transaction as authority metadata. Application semantics remain separate.
package profilestate

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"

	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupstore"
)

const MaxBytes = 16 * 1024 * 1024
const ChunkBytes = 512 * 1024
const manifestKey = "application:state"
const prefix = manifestKey + ":"

type manifest struct {
	Domain string `json:"domain"`
	Bytes  int    `json:"bytes"`
	Chunks int    `json:"chunks"`
	Digest string `json:"digest"`
}
type State struct {
	Bytes  []byte
	Digest string
}

func partKey(i int) string           { return fmt.Sprintf("%s%04d", prefix, i) }
func integrity(message string) error { return fmt.Errorf("%w: %s", groupstore.ErrIntegrity, message) }
func decodeManifest(data []byte) (manifest, error) {
	var m manifest
	if _, err := core.DecodeJSON(data, 1024); err != nil {
		return m, integrity("manifesto privado ilegível")
	}
	if err := json.Unmarshal(data, &m); err != nil {
		return m, integrity("manifesto privado inválido")
	}
	encoded, err := core.Canonical(m)
	if err != nil || !bytes.Equal(encoded, data) || m.Domain != "relayloom/profile-state/1" || m.Bytes < 1 || m.Bytes > MaxBytes || m.Chunks != (m.Bytes+ChunkBytes-1)/ChunkBytes || !core.ValidAddress(m.Digest) {
		return m, integrity("limites ou campos do manifesto privado inválidos")
	}
	return m, nil
}
func Read(tx *groupstore.Tx) (*State, error) {
	data, exists, err := tx.Get(manifestKey)
	if err != nil {
		return nil, err
	}
	keys, err := tx.Keys(prefix)
	if err != nil {
		return nil, err
	}
	if !exists {
		if len(keys) != 0 {
			return nil, integrity("blocos privados sem manifesto")
		}
		return nil, nil
	}
	m, err := decodeManifest(data)
	if err != nil {
		return nil, err
	}
	if len(keys) != m.Chunks {
		return nil, integrity("conjunto de blocos privados incompleto")
	}
	result := make([]byte, 0, m.Bytes)
	for i, key := range keys {
		if key != partKey(i) {
			return nil, integrity("nome de bloco privado inválido")
		}
		piece, exists, err := tx.Get(key)
		if err != nil {
			return nil, err
		}
		if !exists || len(piece) != min(ChunkBytes, m.Bytes-i*ChunkBytes) {
			return nil, integrity("bloco privado truncado")
		}
		result = append(result, piece...)
	}
	if core.Hash(result) != m.Digest {
		return nil, integrity("hash do estado privado não coincide")
	}
	return &State{result, m.Digest}, nil
}

// Write authenticates the previous blob and compares its digest before writing.
// Callers must wait for the outer SQLite commit before emitting side effects.
func Write(tx *groupstore.Tx, data []byte, expectedDigest *string) (string, error) {
	if len(data) < 1 || len(data) > MaxBytes {
		return "", errors.New("estado privado fora dos limites")
	}
	value, err := core.DecodeJSON(data, MaxBytes)
	if err != nil {
		return "", err
	}
	encoded, err := core.Canonical(value)
	if err != nil || !bytes.Equal(encoded, data) {
		return "", errors.New("estado privado não canónico")
	}
	previous, err := Read(tx)
	if err != nil {
		return "", err
	}
	if (previous == nil) != (expectedDigest == nil) || (previous != nil && previous.Digest != *expectedDigest) {
		return "", errors.New("o estado privado mudou; volte a lê-lo antes de guardar")
	}
	digest := core.Hash(data)
	if previous != nil && previous.Digest == digest {
		return digest, nil
	}
	chunks := (len(data) + ChunkBytes - 1) / ChunkBytes
	for i := 0; i < chunks; i++ {
		if err = tx.Put(partKey(i), data[i*ChunkBytes:min(len(data), (i+1)*ChunkBytes)], groupstore.Data); err != nil {
			return "", err
		}
	}
	keys, err := tx.Keys(prefix)
	if err != nil {
		return "", err
	}
	for _, key := range keys[chunks:] {
		if err = tx.Delete(key); err != nil {
			return "", err
		}
	}
	index, err := core.Canonical(manifest{"relayloom/profile-state/1", len(data), chunks, digest})
	if err != nil {
		return "", err
	}
	if err = tx.Put(manifestKey, index, groupstore.Data); err != nil {
		return "", err
	}
	return digest, nil
}

File: packages/profile/src/state.ts
import { canonical, hash } from "../../core/src/index.js";
import {
  RegistryIntegrityError,
  type RegistryTransaction,
} from "../../groups/src/storage.js";

const MANIFEST = "application:state";
const PREFIX = MANIFEST + ":";
export const PROFILE_STATE_LIMITS = Object.freeze({
  bytes: 16 * 1024 ** 2,
  chunkBytes: 512 * 1024,
  chunks: 32,
});
interface StateManifest {
  domain: "relayloom/profile-state/1";
  bytes: number;
  chunks: number;
  digest: string;
}
export interface ProfileStateBytes {
  bytes: Buffer;
  digest: string;
}
const partKey = (i: number) => PREFIX + String(i).padStart(4, "0");
function insist(condition: unknown, message: string): asserts condition {
  if (!condition) throw new RegistryIntegrityError(message);
}
function manifest(data: Buffer): StateManifest {
  try {
    const value = JSON.parse(data.toString("utf8"));
    insist(
      data.length <= 1024 &&
        value &&
        typeof value === "object" &&
        !Array.isArray(value) &&
        Object.keys(value).sort().join(",") === "bytes,chunks,digest,domain" &&
        Buffer.from(canonical(value)).equals(data),
      "Manifesto privado inválido",
    );
    insist(
      value.domain === "relayloom/profile-state/1" &&
        Number.isSafeInteger(value.bytes) &&
        value.bytes > 0 &&
        value.bytes <= PROFILE_STATE_LIMITS.bytes &&
        Number.isSafeInteger(value.chunks) &&
        value.chunks ===
          Math.ceil(value.bytes / PROFILE_STATE_LIMITS.chunkBytes) &&
        typeof value.digest === "string" &&
        /^[a-f0-9]{64}$/.test(value.digest),
      "Limites do manifesto privado inválidos",
    );
    return value;
  } catch (error) {
    if (error instanceof RegistryIntegrityError) throw error;
    throw new RegistryIntegrityError("Manifesto privado ilegível");
  }
}
/** Authenticates the complete bounded blob. Application semantic validation is
 * still required; this function never treats opaque bytes as group authority. */
export function readProfileState(
  tx: RegistryTransaction,
): ProfileStateBytes | null {
  const data = tx.get(MANIFEST),
    keys = tx.keys(PREFIX);
  if (!data) {
    insist(keys.length === 0, "Blocos privados sem manifesto");
    return null;
  }
  const m = manifest(data);
  insist(
    keys.length === m.chunks && keys.every((key, i) => key === partKey(i)),
    "Conjunto de blocos privados incompleto",
  );
  const pieces = keys.map((key, i) => {
    const piece = tx.get(key);
    insist(
      piece &&
        piece.length ===
          Math.min(
            PROFILE_STATE_LIMITS.chunkBytes,
            m.bytes - i * PROFILE_STATE_LIMITS.chunkBytes,
          ),
      "Bloco privado truncado",
    );
    return piece;
  });
  const bytes = Buffer.concat(pieces, m.bytes);
  insist(hash(bytes) === m.digest, "Hash do estado privado não coincide");
  return { bytes, digest: m.digest };
}
/** The caller's transaction commits chunks, manifest and other metadata together.
 * A returned digest is not permission to display/send before that commit returns. */
export function writeProfileState(
  tx: RegistryTransaction,
  bytes: Buffer,
  expectedDigest: string | null,
): string {
  if (
    !Buffer.isBuffer(bytes) ||
    bytes.length < 1 ||
    bytes.length > PROFILE_STATE_LIMITS.bytes
  )
    throw new Error("Estado privado fora dos limites");
  const value = JSON.parse(bytes.toString("utf8"));
  if (!Buffer.from(canonical(value)).equals(bytes))
    throw new Error("Estado privado não canónico");
  const previous = readProfileState(tx);
  if ((previous?.digest ?? null) !== expectedDigest)
    throw new Error("O estado privado mudou; volte a lê-lo antes de guardar");
  const digest = hash(bytes);
  if (previous?.digest === digest) return digest;
  const chunks = Math.ceil(bytes.length / PROFILE_STATE_LIMITS.chunkBytes);
  for (let i = 0; i < chunks; i++)
    tx.put(
      partKey(i),
      bytes.subarray(
        i * PROFILE_STATE_LIMITS.chunkBytes,
        (i + 1) * PROFILE_STATE_LIMITS.chunkBytes,
      ),
    );
  for (const key of tx.keys(PREFIX).slice(chunks)) tx.delete(key);
  tx.put(
    MANIFEST,
    Buffer.from(
      canonical({
        domain: "relayloom/profile-state/1",
        bytes: bytes.length,
        chunks,
        digest,
      }),
    ),
  );
  return digest;
}
